### ISSUE TITLE
Document ability to use OR conditional in #states

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#4730]
When using Form API `#states` to make fields visible/hidden, etc. based on the value(s) of other fields, you can use an `OR` conditional to, say, show this field when this other field's value is 'x' OR that other field's value is 'y'. However this ability to use `OR` is un-documented in Backdrop. Drupal has a change record for it though: https://www.drupal.org/node/1464758 (which works in Backdrop too).

Rather than having to consult the Drupal CR each time I implement this (because I can never remember the right syntax), let's document it in [backdrop_process_states()](https://github.com/backdrop/backdrop/blob/1.x/core/includes/common.inc#L5199)

**Comment by BWPanda** [comment]
PR: https://github.com/backdrop/backdrop/pull/3461

**Comment by klonos** [comment]
Thanks @BWPanda 🙏 ...that example looks good, but should we be explaining things a bit further up in that docblock, where the rest of the generic pseudo-code examples are? Here's what I think should end up in https://api.backdropcms.org/api/backdrop/1/search/backdrop_process_states

> #states is an associative array in the form of:
>
> ```php
> '#states' => array(
>   STATE1 => CONDITIONS_ARRAY1,
>   STATE2 => CONDITIONS_ARRAY2,
>   ...
> )
> ```
>
> Each key is the name of a state to apply to the element, such as 'visible' or
> 'disabled'. Each value is a list of conditions that denote when the state
> should be applied.
>
> Multiple conditions may be specified for a single state, and the default
> conditional is AND:
> 
> ```php
> '#states' => array(
>   STATE1 => array(
>     JQUERY_SELECTOR1 => CONDITION1,
>     // This syntax means:
>     // "the condition above needs to be true, AND the one below...".
>     JQUERY_SELECTOR2 => CONDITION2,
>     // "...AND the one below".
>     JQUERY_SELECTOR3 => CONDITION3,
>     // "...AND so on".
>     ...
>   ),
>   STATE2 => CONDITIONS_ARRAY2,
>   ...
> )
> ```
> 
> For OR conditions, use a non-associative array by wrapping each group of
> conditions in the array() function:
> 
> 
> ```php
> '#states' => array(
>   STATE1 => array(
>     array(
>       JQUERY_SELECTOR1 => CONDITIONS_ARRAY1,
>     ),
>     // This syntax means:
>     // "any or both of the conditions may be true (at least one); the one
>     // above, OR the one below".
>     array(
>       JQUERY_SELECTOR2 => CONDITIONS_ARRAY2,
>     ),
>   ),
>   STATE2 => CONDITIONS_ARRAY2,
>   ...
> )
> ```
> 
> For XOR conditions, a similar syntax as OR can be used, with a non-associative
> array; but while the OR operator is implied when using a non-associative
> array, the XOR operator needs to > be explicitly defined, with an array item
> with the string 'xor': 
> 
> ```php
> '#states' => array(
>   STATE1 => array(
>     array(
>       JQUERY_SELECTOR1 => CONDITIONS_ARRAY1,
>     ),
>     // This syntax means "one or the other condition is true, BUT NOT BOTH".
>     'xor',
>     array(
>       JQUERY_SELECTOR2 => CONDITIONS_ARRAY2,
>     ),
>   ),
>   STATE2 => CONDITIONS_ARRAY2,
>   ...
> )
> ```

Then the real-code example snippets can follow.

**Comment by BWPanda** [comment]
Thanks @klonos. I used most of your suggestions, with a few changes (as some of your syntax wasn't quite correct).

**Comment by klonos** [comment]
> as some of your syntax wasn't quite correct

Not sure about that. Consider this bit of docblock re the OR condition, taken from your PR, and edited to remove the inline comment + simplified to a single state:


```php
 * @code
 * '#states' => array(
 *   STATE1 => array(
 *     array(
 *       array(JQUERY_SELECTOR1 => CONDITION1),
 *       array(JQUERY_SELECTOR2 => CONDITION2),
 *     ),
 *   ),
 * )
```

Notice how you seem to have added an extra level of nesting with an additional array, which is not correct. In other words, you are suggesting that an actual OR example would be:

```php
  '#states' => array(
    'visible' => array(
      array(
        array(
          ':input[name="name"]' => array('filled' => TRUE),
        ),
        array(
          ':input[name="anonymous"]' => array('checked' => TRUE),
        ),
      ),
    ),
  ),
```

Whereas this is what it should look like instead:

```php
  '#states' => array(
    'visible' => array(
      array(
        ':input[name="name"]' => array('filled' => TRUE),
      ),
      array(
        ':input[name="anonymous"]' => array('checked' => TRUE),
      ),
    ),
  ),
```

For comparison, the respective XOR should be:

```php
  '#states' => array(
    'visible' => array(
      array(
        ':input[name="name"]' => array('filled' => TRUE),
      ),
      'xor',
      array(
        ':input[name="anonymous"]' => array('checked' => TRUE),
      ),
    ),
  ),
```

...and AND would be:

```php
  '#states' => array(
    'visible' => array(
      ':input[name="name"]' => array('filled' => TRUE),
      ':input[name="anonymous"]' => array('checked' => TRUE),
    ),
  ),
```

**Comment by BWPanda** [comment]
> Notice how you seem to have added an extra level of nesting with an additional array, which is not correct.

That *is* actually correct and is entirely my point.

> In other words, you are suggesting that an actual OR example would be:

Exactly.

Please see the Drupal changelog for the syntax/examples.

The important part, I believe (it's not explained in the changelog but I think I worked out the reasoning), is that ANDs have the selector as the array key and hence you can't have two conditions based on the same selector (one would override the other). For ANDs this is fine though (it doesn't make sense to have two conditions where name value is filled AND name is also checked).

For ORs though, it's very likely you want to have two conditions with the same selector (e.g. name has value 'x' OR name has value 'y'). In this case, you add a top-level array to group the ORs together (to distinguish from ANDs), then you wrap each selector in its own second array so the the top-level array key becomes an automatically-assigned integer, allowing you to specify the same selector multiple times without them overriding each other.

I hope that explains it better...

**Comment by klonos** [comment]
> Please see the Drupal changelog for the syntax/examples.

I see. That example in the d.org change record is wrong, in that it combines multiple examples within a single one, using the default OR. It should be fixed, but more on that later. My go-to article for many years has been this: https://www.lullabot.com/articles/form-api-states (you will find individual examples for AND/OR/XOR there, which are correct, and work as expected)

## What's wrong with example in the d.org change record

It should be split into 3 separate examples. So this:

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
    // dependee_1 has value ON
      '[name="dependee_1"]' => array('value' => 'ON'),
      array(
      // At least one of dependee_2 or dependee_3 has value ON
        array('[name="dependee_2"]' => array('value' => 'ON')),
        array('[name="dependee_3"]' => array('value' => 'ON')),
      ),
      array(
      // Only one of dependee_4 or dependee_5 can have value ON
        array('[name="dependee_4"]' => array('value' => 'ON')),
        'xor',
        // The field should be disabled when Select #1 has value ON; at least one
        // of Select #2 and Select #3 has value ON; one but only one of Select #4
        // and Select #5 has value ON.
        array('[name="dependee_5"]' => array('value' => 'ON')),
      ),
    )
  );
```

...should be:

Simple state example, with a single condition

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
      '[name="dependee_1"]' => array('value' => 'ON'),
    )
  );
```

Simple OR example:

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
      // At least one of dependee_2 or dependee_3 has value ON.
      array('[name="dependee_2"]' => array('value' => 'ON')),
      array('[name="dependee_3"]' => array('value' => 'ON')),
    ),
  );
```

...which btw becomes an AND if you omit the arrays in the 2 conditions:

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
      // Both dependee_2 and dependee_3 has value ON.
      '[name="dependee_2"]' => array('value' => 'ON'),
      '[name="dependee_3"]' => array('value' => 'ON'),
    ),
  );
```

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
      // Only one of dependee_4 or dependee_5 can have value ON
      array('[name="dependee_4"]' => array('value' => 'ON')),
      'xor',
      // The field should be disabled when Select #1 has value ON; at least one
      // of Select #2 and Select #3 has value ON; one but only one of Select #4
      // and Select #5 has value ON.
      array('[name="dependee_5"]' => array('value' => 'ON')),
    ),
  );
```

In other words, the d.org example has this:

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
      // Simple condition.
      array(
        // OR example.
      ),
      array(
        // XOR example.
      ),
    )
  );
```

...which basically if executed as a whole, means:

```php
  $form['dependent_1']['#states'] = array(
    'disabled' => array(
      // Simple condition.
      // OR the OR example.
      // OR the XOR example.
    )
  );
```

To my point: additional arrays within conditions (nested or not) means OR.

Also see the comment by that user in that change record, which is trying to make things work with the additional nested arrays, and of course things fail.

**Comment by klonos** [comment]
...see my comment in https://www.drupal.org/node/1464758#comment-13949961 with all my simplified examples, which make sense, and most importantly: they work! 😉

**Comment by BWPanda** [comment]
Wow, what a mess! Weird to think that a Lullabot article has more up-to-date and accurate information than a Drupal changelog... (no offence @quicksketch).

**Comment by BWPanda** [comment]
> and most importantly: they work!

FTR, my code works too 😝

**Comment by BWPanda** [comment]
Updated PR again.

**Comment by klonos** [comment]
> Updated PR again.

Thanks @BWPanda 🙏 

> FTR, my code works too 😝

Agreed, but it is complicated, and it only works because it is basically set to "expect" another nested "level 1" array to be added 😅

**Comment by klonos** [comment]
Here's a very simplified, and straight-forward real-code example, which shows all possible AND/OR/XOR conditions, and will run if a developer were to add it somewhere for testing purposes:

```php
$form['test_states'] = array(
  '#type' => 'checkboxes',
  '#title' => t('Test #states'),
  '#options' => backdrop_map_assoc(array('A', 'B')),
  '#description' => t('Click either or both checkboxes above, and see what happens to the states-controlled checkboxes below.'),
);
$form['test_and'] = array(
  '#type' => 'checkbox',
  '#title' => t('AND works (both A and B are checked)'),
  '#disabled' => TRUE,
  '#states' => array(
    'checked' => array(
      ':input[name="test_states[A]"]' => array('checked' => TRUE),
      ':input[name="test_states[B]"]' => array('checked' => TRUE),
    ),
  ),
);
$form['test_or'] = array(
  '#type' => 'checkbox',
  '#title' => t('OR works (at least one of A or B is checked)'),
  '#disabled' => TRUE,
  '#states' => array(
    'checked' => array(
      array(
        ':input[name="test_states[A]"]' => array('checked' => TRUE),
      ),
      'or', // This is the default operator, so it may be omitted.
      array(
        ':input[name="test_states[B]"]' => array('checked' => TRUE),
      ),
    ),
  ),
);
$form['test_xor'] = array(
  '#type' => 'checkbox',
  '#title' => t('XOR works (one of A or B is checked, but not both)'),
  '#disabled' => TRUE,
  '#states' => array(
    'checked' => array(
      array(
        ':input[name="test_states[A]"]' => array('checked' => TRUE),
      ),
      'xor',
      array(
        ':input[name="test_states[B]"]' => array('checked' => TRUE),
      ),
    ),
  ),
);
```

@BWPanda do you think that it'd be worth adding such a "complete" code example to the docblock? (so that it ends up on the API page) ...or bit too much?

**Comment by klonos** [comment]
...either way, this is RTBC 👍

**Comment by BWPanda** [comment]
I think what we've added so far is a huge step forward. Let's commit this and then see what (if any) feedback we get in future.

**Comment by klonos** [comment]
Sure @BWPanda 👍 ...I might add that code snippet as a blog post on b.org, or as a topic on our Forum then.

**Comment by klonos** [comment]
FTR, blog post drafted here: https://backdropcms.org/news/states-and-or-xor-conditionals-explained

**Comment by quicksketch** [comment]
Thanks @BWPanda and @klonos! I LOOOOOVE documentation fixes and enhancements like this! `#states` is so darn tricky and I have to search around every time I need complex ones. This is definitely a good thing to put the documentation where it can be found. Merged https://github.com/backdrop/backdrop/pull/3461 into 1.x and 1.18.x.
